Port ISIS Sans to use Qt adaptor

DavidFair opened the pull request:
**Description of work.**

    Switch SANS Components to Qt/Threading adaptor
    
    Switches all the SANS components away from a bespoke threading solution
    to the generic Qt Threading adaptor.
    
    This brings a number of advantages:
    - Stack traces without execute() steps included
    - Increased responsiveness (subjectively)
    - Better tooling support

**Note** The methods in files belonging to the `_async` module were cut and paste from the existing models / presenters (where they incorrectly resided.
As this is a cut and paste a reviewer is welcome to review each method, but does not need to for new async files.

**To test:**
- TODO - use https://developer.mantidproject.org/Testing/SANSGUI/SANSGUITests.html
- 

Fixes #27895 . 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Switch SANS Components to Qt/Threading adaptor

Switches all the SANS components away from a bespoke threading solution
to the generic Qt Threading adaptor.

This brings a number of advantages:
- Stack traces without execute() steps included
- Increased responsiveness (subjectively)
- Better tooling support

DavidFair committed:
Modify adaptor to call cb in unit test mode

This allows tests to actually verify the output, without having to
resort to tricks like running Qt message pumps.
It also mirrors the output in a synchronous way to a mocked callback
handler bringing the tests more in-line with the actual usage

DavidFair committed:
Update async diag worker to return results for testing

The presenter for the diagnostic page does not do anything with the
result so this ended up missed. Instead we add this method for testing
to ensure that if we did end up using it in the view, it would work as
expected.

DavidFair committed:
Revert passing single item of GUIState

Reverts passing a single item of GUIState back to a list of length 1,
that consists of all states. This is to prevent us refactoring the
underlying algorithm that assumes we will be doing this.

Pasarus reviewed (changes requested):
Largely everything seems functional as expected, there seems to be some discrepancies however, notably the HAB workspaces and plots being present when they are explicitly left out of the instructions, for 1D and 2D processing, hopefully this is not (my) user error.

Also these buttons don't exist and thus can't be checked in the beam centre tab:
`Both the Update main-detector and Update Hab checkboxes should be checked and enabled.`

When performing the following steps of sum runs I get a very long message in the console:
```
9. Change the first column of both rows to LOQ74044-add.
10. Click Process All.
```
The Long message:
```
Missing detector at 16734
Missing detector at 16735
Missing detector at 17084
Missing detector at 17085
Missing detector at 17434
Missing detector at 17435
Missing detector at 17784
Missing detector at 17785
Missing detector at 17786
Missing detector at 17787
Missing detector at 17788
Missing detector at 17789
Missing detector at 17790
Missing detector at 17791
Error in execution of algorithm SANSLoad:
No registered converter was able to produce a C++ rvalue of type std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > from this Python object of type NoneType
  at line 141 in '/home/sam/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSLoad.py'
  caused by line 336 in '/home/sam/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSLoad.py'
  caused by line 44 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_sans_instrument_component.py'
  caused by line 363 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 587 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 49 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 122 in '/home/sam/mantid/scripts/SANS/sans/common/general_functions.py'
  ```
 Similar when checking the diagnostic instructions and clicking the Integral buttons (All of them):
 ```
 SANSLoad started
Error in execution of algorithm SANSLoad:
No registered converter was able to produce a C++ rvalue of type std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > from this Python object of type NoneType
  at line 141 in '/home/sam/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSLoad.py'
  caused by line 336 in '/home/sam/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSLoad.py'
  caused by line 44 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_sans_instrument_component.py'
  caused by line 363 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 587 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 49 in '/home/sam/mantid/scripts/SANS/sans/algorithm_detail/move_workspaces.py'
  caused by line 122 in '/home/sam/mantid/scripts/SANS/sans/common/general_functions.py'
```

Finally a probably unrelated issue with missing tooltips on:
- Settings -> Mask -> Mask File input
- Settings -> General, Scale, Event Slice, Sample -> Slices (text only, box is fine, tooltip needs duplication for here)
- Settings -> Q, Wavelength, Detector Limits -> Wavelength (Whole box)
- Settings -> Q, Wavelength, Detector Limits -> Q Limits (Whole box)
- Settings -> Q, Wavelength, Detector Limits -> Account for gravity (Whole box)
- Settings -> Q, Wavelength, Detector Limits -> Q resolution (Whole box)
- Settings -> State Diagnostic (The tab)
- Beam Centre -> Options -> Radius Limits (text boxes and QLabels)
- Beam Centre -> Options -> Max iterations (QLabel)
- Beam Centre -> Options -> General Options (checkbox and label)
- Beam Centre -> Options -> Directions (checkbox and label)
- Beam Centre -> Run (I would be remiss if this button needed a tooltip but hey it's listed to check them)
- Sum runs (All of this tab except the Select Save Directory button and QLabel)
- Diagnostic Page (All)

DavidFair commented on scripts/SANS/sans/gui_logic/models/async_workers/sans_run_tab_async.py:
```diff
@@ -0,0 +1,107 @@
+                continue
+
+            assert len(states) == 1
+            # XXX: Replace this when get_states_func stops returning a dict of len 1
```
In this case no, since we already unit test this but it's more a sanity check to make sure no obscure part of the codebase somehow triggers pushes through 2 states.

This check has been here for a while and thus far myself or the users doing very obscure things have hit it. So now the plan is to pass a single object through rather than a dict of length one

DavidFair commented on scripts/SANS/sans/gui_logic/models/async_workers/sans_run_tab_async.py:
```diff
@@ -0,0 +1,107 @@
+                continue
+
+            assert len(states) == 1
+            # XXX: Replace this when get_states_func stops returning a dict of len 1
```
^

DavidFair commented on scripts/SANS/sans/gui_logic/models/async_workers/sans_run_tab_async.py:
```diff
@@ -0,0 +1,107 @@
+    @staticmethod
+    def _mark_row_error(row: RowEntries, error: Exception):
+        # We manually have to extract out the traceback, since going to a str for Qt signals will strip this
+        # self._logger.error(''.join(traceback.format_tb(e.__traceback__)))
```
This one can go away now that we get native stave traces :+1:

DavidFair commented on scripts/SANS/sans/gui_logic/models/sum_runs_model.py:
```diff
@@ -5,30 +5,24 @@
+class SumRunsModel(IQtAsync):
+    def __init__(self, view=None):
+        super().__init__()
+        # TODO this should be in presenter, not the model
```
It depends:
- If your happy for a much larger diff I can move the tests and slots to the presenter 
- Equally I won't create an issue if we want to leave this, because I'll pick it up next time I touch the file (continuous refactoring rather then dedicated)

I'm happy to do either?

DavidFair replied:
I'm going to leave this for the future when I next touch anything related to Sum Runs

DavidFair committed:
Address review comments

This includes:
- Renaming test files to use test_* convention
- Some dead code elimination
- Clarification on why some "dead code" is required

DavidFair commented:
@Pasarus Could you look at this, I've resolved the merge conflicts again

martyngigg commented:
@DavidFair I'm afraid there are merge conflicts again after I merged another PR. Let me know as soon as they are fixed and I'll set it to auto merge.

DavidFair commented:
@martyngigg Thanks, I've also manually re-tested this too as it was a nasty merge conflict
